File: src/leetcode75/ValidParentheses.java
import java.util.Stack;

public class ValidParentheses {
    public static boolean isValidParentheses(String s) {
        Stack<Character> stack = new Stack<>();
        for (int i = 0; i < s.length(); i++) {
            Character c = s.charAt(i);
            if (c == '(' || c == '{' || c == '[') {
                stack.push(c);
            } else {
                if (stack.isEmpty()) {
                    return false;
                }
                Character pop = stack.pop();
                if ((c == ')' && pop != '(') || (c == '}' && pop != '{') || (c == ']' && pop != '[')) {
                    return false;
                }
            }
        }

        return stack.isEmpty();
    }

    public static void main(String[] args) {
        String test1 = "()";
        String test2 = "()[]{}";
        String test3 = "(]";

        System.out.println(isValidParentheses(test1));
        System.out.println(isValidParentheses(test2));
        System.out.println(isValidParentheses(test3));
    }
}

File: src/main/python/validparentheses.py
class Solution(object):
    def valid_parentheses(self, s):
        """
        :type s: str
        :rtype: bool
        """
        if len(s) == 0:
            return True
        if len(s) % 2 != 0:
            return False

        mapping = {")": "(", "}": "{", "]": "["}

        stack = []
        for c in s:
            if c == '(':
                stack.append(c)
            elif c == '{':
                stack.append(c)
            elif c == '[':
                stack.append(c)
            else:
                if len(stack) == 0:
                    return False
                if stack.pop() != mapping[c]:
                    return False
        return len(stack) == 0

if __name__ == '__main__':
    s = Solution()
    print(s.valid_parentheses("()"))
    print(s.valid_parentheses("()[]{}"))
    print(s.valid_parentheses("(]"))
    print(s.valid_parentheses("([)]"))
    print(s.valid_parentheses("{[]}"))
